armi/reactor/blocks.py

```diff
@@ -2167,6 +2197,26 @@ def _rotatePins(self, rotNum, justCompute=False):
 
         return rotateIndexLookup
 
+    def _rotatePinParameters(self, rotNum: int):
+        """Rotate parameters assigned on pins."""
+        params = self.p.where(lambda pd: pd.atLocation(ParamLocation.CHILDREN))
```

Comment:
It seems like we are fundamentally redefining the way pin-level data is stored in ARMI here.

Up to this point, parameter data that exists at `ParamLocation.CHILDREN` has always been associated with pin indices. For some parameter defined on pins, e.g., `linPowByPin`, the data never gets reshuffled after a rotation; only `pinLocation` gets rotated. `linyPowByPin[n]` will always contain the data for the `n`th pin. If you want to know where the `n`th pin is currently located, you would use `_rotatePins` and/or the `pinLocation` parameter to find out.

For example, see the method `setPinPowers`:

https://github.com/terrapower/armi/blob/96a5c88d6167cee989378599b6d341c351c291bd/armi/reactor/blocks.py#L1983-L1990

An array of data is passed in that represents a distribution of pin powers indexed by pin location. The `pinLocation` is mapped to a `pinNum` before being stored in the `linPowByPin*` parameter, using `pinNum` as an index.

If we start rotating `pinLocation` and also `linPowByPin*` and other pin-level parameters, then there's a risk of double-counting rotations.

Fundamentally, I think any parameter that is `ParamLocation.CHILDREN` should not have any information about how much the block has been rotated baked into it (other than `pinLocation`). The rotations should be tracked by the `pinLocation` parameter, while the other parameters remain invariant to rotation.

Reply:
Okay, I've been digging through the code base and internal projects to better understand rotation. What @mgjarrett brings up in

> For some parameter defined on pins, e.g., linPowByPin, the data never gets reshuffled after a rotation; only pinLocation gets rotated. linyPowByPin[n] will always contain the data for the nth pin. If you want to know where the nth pin is currently located, you would use _rotatePins and/or the pinLocation parameter to find out

matches what I'm seeing. I would probably lean towards calling `pinLocation` something like `pinIndex` because it is _really_ a mapping from values in an (potentially unrotated) data array where that pin's data should actually live in parameters but that's not the point.

This brings up a weird point/implication. Things calling `setPinPowers` and, similarly, [`Block.setPinMgFluxes`](https://github.com/terrapower/armi/blob/f1f3dbea05c2564404a03d8e053f72c2af15daa9/armi/reactor/blocks.py#L377-L383) don't use any of the spatial information from the block. Data are rotated in the `setPin*` calls. By further extension, any pin-like parameter would need a similar setter, e.g., `setPercentBuByPin`. 

If they attempt to pull spatial information from [`Block.getPinCoords`](https://github.com/terrapower/armi/blob/f1f3dbea05c2564404a03d8e053f72c2af15daa9/armi/reactor/blocks.py#L1589-L1595), that reflects the unrotated state. Which is probably something that should be rotated so the location of `pin[n]` found in `Block.getPinCoords()[n]` reflects where that pin exists in the block before and after rotation.

But, doing that would likely mean we have another double rotation problem that @mgjarrett pointed out
1. Pin locations from `getPinCoords` are rotated.
2. Data obtained using the current pin locations.
3. Data are "re-rotated" mapping through `b.p.pinLocation`

## Decisions
1. Do we want to update the spatial location of pins through rotation as reported by `Block.getPinCoords`?
   - I'm leaning towards yes because you'd expect XY coordinates to change through rotation
 2. Do we want block spatial parameters to reflect the current orientation of the block (like this PR proposed), or the un-rotated state of the block (like we currently do)
   - I'm leaning towards yes partially because sunk cost fallacy (I've written code to do that already) but it also feels odd that you could rotate a block, look at some parameter data, and not _really_ know where that data exists in the block
   - Docs would help
   - We'd also need logic to handle parameter rotation when pin-like data are set for every pin-like parameter